File: ts/src/store.ts
import { ChainIntegrityError, EventNotFoundError } from "./errors.js";
import { Event } from "./event.js";
import { EventId, Hash, Option } from "./types.js";

export interface ChainVerification {
  valid: boolean;
  length: number;
}

export interface Store {
  append(event: Event): Event;
  get(eventId: EventId): Event;
  head(): Option<Event>;
  count(): number;
  verifyChain(): ChainVerification;
  close(): void;
}

export class InMemoryStore implements Store {
  private readonly events: Event[] = [];
  private readonly index = new Map<string, number>();

  append(event: Event): Event {
    if (this.events.length > 0) {
      const last = this.events[this.events.length - 1];
      if (event.prevHash.value !== last.hash.value) {
        throw new ChainIntegrityError(
          this.events.length,
          `prev_hash ${event.prevHash.value} != head hash ${last.hash.value}`,
        );
      }
    }
    this.events.push(event);
    this.index.set(event.id.value, this.events.length - 1);
    return event;
  }

  get(eventId: EventId): Event {
    const pos = this.index.get(eventId.value);
    if (pos === undefined) throw new EventNotFoundError(eventId.value);
    return this.events[pos];
  }

  head(): Option<Event> {
    if (this.events.length === 0) return Option.none();
    return Option.some(this.events[this.events.length - 1]);
  }

  count(): number {
    return this.events.length;
  }

  verifyChain(): ChainVerification {
    for (let i = 1; i < this.events.length; i++) {
      if (this.events[i - 1].hash.value !== this.events[i].prevHash.value) {
        return { valid: false, length: i };
      }
    }
    return { valid: true, length: this.events.length };
  }

  recent(limit: number): Event[] {
    return this.events.slice(-limit).reverse();
  }

  close(): void {}
}

File: go/pkg/store/store.go
package store

import (
	"github.com/transpara-ai/eventgraph/go/pkg/event"
	"github.com/transpara-ai/eventgraph/go/pkg/types"
)

// Store is the event and edge persistence interface.
// Implementations must be safe for concurrent access.
// Append serialises hash chain writes internally.
type Store interface {
	// Append persists a pre-built Event from EventFactory.
	// Validates PrevHash matches chain head, recomputes and verifies Hash.
	// Idempotent: if the same ID already exists, returns the stored event without error.
	Append(ev event.Event) (event.Event, error)

	// Get retrieves an event by ID.
	Get(id types.EventID) (event.Event, error)

	// Head returns the most recent event, or None if the store is empty.
	Head() (types.Option[event.Event], error)

	// Recent returns the most recent events.
	Recent(limit int, after types.Option[types.Cursor]) (types.Page[event.Event], error)

	// ByType returns events of a specific type.
	ByType(eventType types.EventType, limit int, after types.Option[types.Cursor]) (types.Page[event.Event], error)

	// BySource returns events from a specific actor.
	BySource(source types.ActorID, limit int, after types.Option[types.Cursor]) (types.Page[event.Event], error)

	// ByConversation returns events in a conversation thread.
	ByConversation(id types.ConversationID, limit int, after types.Option[types.Cursor]) (types.Page[event.Event], error)

	// Since returns events after a specific event ID.
	Since(afterID types.EventID, limit int) (types.Page[event.Event], error)

	// Ancestors returns causal ancestors up to maxDepth.
	Ancestors(id types.EventID, maxDepth int) ([]event.Event, error)

	// Descendants returns causal descendants up to maxDepth.
	Descendants(id types.EventID, maxDepth int) ([]event.Event, error)

	// EdgesFrom returns edges originating from an actor of a specific type.
	EdgesFrom(id types.ActorID, edgeType event.EdgeType) ([]event.Edge, error)

	// EdgesTo returns edges pointing to an actor of a specific type.
	EdgesTo(id types.ActorID, edgeType event.EdgeType) ([]event.Edge, error)

	// EdgeBetween returns the edge between two actors of a specific type, if any.
	EdgeBetween(from types.ActorID, to types.ActorID, edgeType event.EdgeType) (types.Option[event.Edge], error)

	// Count returns the total number of events.
	Count() (int, error)

	// VerifyChain verifies the hash chain integrity.
	VerifyChain() (event.ChainVerifiedContent, error)

	// Close releases resources.
	Close() error
}

// IIdentity provides signing and verification for events.
type IIdentity interface {
	// SystemURI returns this system's identity URI.
	SystemURI() types.SystemURI

	// PublicKey returns this system's public key.
	PublicKey() types.PublicKey

	// Sign signs the given data, returning a signature.
	Sign(data []byte) (types.Signature, error)

	// Verify checks a signature against a public key and data.
	Verify(publicKey types.PublicKey, data []byte, signature types.Signature) (bool, error)
}
